### ISSUE TITLE
Enable endpoint customization for each integration

### ISSUE THREAD

**Issue by maciejwalkowiak** [spring-cloud/spring-cloud-aws#570]
Each AWS integration should provide a way to specify custom endpoint url via `application.properties`. For example:

```
cloud.aws.sqs.endpoint-url=http://localhost:4567
```

This would allow to set easily services to use Localstack (#333) for testing or VPC endpoints (#152) in production.

**Comment by DarrenForsythe** [comment]
@maciejwalkowiak what would the prefered impl. of this? Given the relative complexity of how the modules are setup would it be easiest to just add new configuration properties to their relevant modules. 

For example, 

a new `AwsSNSProperties` class in `spring-cloud-aws-messaging` with two fields `endpoint` and `region` that would be applied to the SNS Client?

**Comment by maciejwalkowiak** [comment]
@DarrenForsythe it's still under investigation. One thing is endpoint customization but it also has to play nice with ClientConfiguration customisation and ideally properties should not require changes once we move to AWS SDK v2. 

Properties will be placed in autoconfigure module for sure but quite a lot of changes in case of messaging are required to make it work with messaging module.


Don't jump on it (yet) please.

**Comment by mikaelhg** [comment]
Since currently everyone who uses Spring Cloud AWS will currently have to choose between either not testing their code, or writing a custom solution to this problem, a quick resolution would trump a perfect resolution. Especially since as long as the parameter names stay the same, the underlying implementation can be swapped out for a better one at will.

**Comment by maciejwalkowiak** [comment]
Current idea: create base class all integrations would inherit from:

```
public abstract class AwsServiceProperties {
	private EndpointProperties endpoint;

	public static class EndpointProperties {
		private URI uri;
		private Regions signingRegion;
	}
}
```

Update `AmazonWebserviceClientFactoryBean` to accept `EndpointConfiguration`.

In each integration auto-configuration follow pattern:

```
// create `<service>EndpointConfiguration` bean only if property is present
@Bean(name = "sqsEndpointConfiguration")
@ConditionalOnProperty("cloud.aws.sqs.endpoint")
public EndpointConfiguration sqsEndpointConfiguration(SqsProperties sqsProperties) {
	return new EndpointConfiguration(sqsProperties.getEndpoint().getUri().toString(),
			sqsProperties.getEndpoint().getSigningRegion().getName());
}

// pass this bean as optional with Qualifier to bean factory creating AWS client
@Lazy
@Bean(destroyMethod = "shutdown")
public AmazonSQSBufferedAsyncClient amazonSQS(@Qualifier("sqsEndpointConfiguration") Optional<EndpointConfiguration> endpointConfiguration) throws Exception {
	AmazonWebserviceClientFactoryBean<AmazonSQSAsyncClient> clientFactoryBean = new AmazonWebserviceClientFactoryBean<>(
			AmazonSQSAsyncClient.class, this.awsCredentialsProvider,
			this.regionProvider);
	endpointConfiguration.ifPresent(clientFactoryBean::setEndpointConfiguration);
	clientFactoryBean.afterPropertiesSet();
	return new AmazonSQSBufferedAsyncClient(clientFactoryBean.getObject());
}
```

This is not exactly going to work for Secrets Manager and Parameter Store integration as they are detached from the `autoconfigure` module - meaning they wont be able to extend from `AwsServiceProperties`. Other than that they can follow similar pattern.

All of these should be possible to be done once #636 and #598 are merged.

**Comment by mikaelhg** [comment]
Including the EC2 metadata service endpoint override, a.k.a. `com.amazonaws.sdk.ec2MetadataServiceEndpointOverride`?

**Comment by maciejwalkowiak** [comment]
I had in mind setting endpoint for AWS clients. Do you think it makes sense for EC2 metadata service too since it can be already relatively easily configured through system property?

**Comment by mikaelhg** [comment]
It makes sense to make Spring Cloud applications as easily testable as other Spring applications, from a standard configuration, rather than throwing unexplained exceptions at test execution, and then requiring every user to solve the same issues with testing independently.

**Comment by mikaelhg** [comment]
POC of a workaround while waiting for proper testing tools: https://github.com/mikaelhg/aws-test-conf-poc

**Comment by maciejwalkowiak** [comment]
Closed in #715
